src/notification-banner/NotificationBanner.jsx

```diff
@@ -0,0 +1,103 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import uuidv4 from 'uuid/v4';
+
+import BannerRegistrationContext from './private/BannerRegistrationContext';
+
+const propTypes = {
+  /**
+   * The text and corresponding callback to populate the action button of the banner.
+   */
+  bannerAction: PropTypes.shape({
+    /**
+     * The text to display in the banner button.
+     */
+    text: PropTypes.string,
+    /**
+     * The Callback function triggered when the action button is clicked. No parameters are passed.
+     */
+    onClick: PropTypes.func,
+  }),
+  /**
+   * The message content to display in the banner.
+   */
+  description: PropTypes.node,
+  /**
+   * Callback function triggered when the dismiss button is clicked. The presence of this prop will cause
+   * the dismiss button to be included on the banner. No parameters are passed.
+   */
+  onRequestClose: PropTypes.func,
+  /**
+   * The variant of notification banner to be rendered. This renders the banner with the corresponding header and icon to the
+   * variant concept.
+   */
+  variant: PropTypes.oneOf([
+    'hazard-high',
+    'hazard-medium',
+    'hazard-low',
+    'error',
+    'unsatisfied',
+    'unverified',
+    'custom',
+  ]).isRequired,
+  /**
+   * The pieces to populate a banner when `variant="custom"`.
+   */
+  custom: PropTypes.shape({
+    /**
+     * The keyword used to represent & emphasis the intention of banner description that is being shown to the user.
+     */
+    signalWord: PropTypes.string,
+    /**
+     * The class name used to set the icon as the background image to be used as the icon in the banner.
+     */
+    iconClassName: PropTypes.string,
+  }),
+};
+
+const NotificationBanner = ({
+  bannerAction, custom, description, onRequestClose, variant,
+}) => {
+  /**
+   * A unique identifier is generated for each Banner during construction. This will be used to
+   * uniquely register/unregister the banner with ancestor Banner Managers without requiring consumers to
+   * define unique identifiers themselves.
+   */
+  const bannerRegistration = React.useContext(BannerRegistrationContext);
+
+  React.useEffect(() => {
+    const uuid = uuidv4();
```

Comment:
Did generating the uuid per-registration (vs per instance, stored in a ref) make registration logic simpler? I like how the status provider had two effects, one to handle registration (and re-registration, if values changed) and another that handles unregistration that will only execute once: https://github.com/cerner/terra-application/blob/main/src/application-status-overlay/ApplicationStatusOverlay.jsx#L41-L51

Having to unregister just to register again seems heavy, even if it doesn't have a visual impact. It's also nice to be able to tie an instance of this component to a single identifier for debugging purposes. 🤷‍♂️ Thoughts?

Reply:
> Did generating the uuid per-registration (vs per instance, stored in a ref) make registration logic simpler?  

it didn't affect the registration logic, the linter was complaining and I didn't notice any adverse affect when I moved it. I see your point on only un-registering once. I had logic similar to that at one point and then changed it. I'll work on updating this to be lighter.

Reply:
fixed: https://github.com/cerner/terra-application/pull/45/commits/de735223046eaabc526799ffde4a1caedc8db758#diff-1c1abe24b9b3aaa6ad0c051069883f78R67